portfolio/src/main/java/com/google/sps/servlets/ChartServlet.java

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.cloud.datastore.Datastore;
+import com.google.cloud.datastore.DatastoreOptions;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.EntityQuery.Builder;
+import com.google.cloud.datastore.Query;
+import com.google.cloud.datastore.QueryResults;
+import com.google.cloud.datastore.StructuredQuery;
+import com.google.cloud.datastore.StructuredQuery.PropertyFilter;
+import com.google.common.collect.Lists;
+import com.google.common.io.CharStreams;
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import java.io.IOException;
+import java.text.DateFormat; 
+import java.text.SimpleDateFormat; 
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Iterator;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/chart")
+public class ChartServlet extends HttpServlet {
+
+  /*
+   * Called when a client submits a POST request to the /chart URL
+   * Prepares data about the number of comments each day and submits
+   * it to the client for rendering
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    HashMap<String, DayComments> numCommentsOnDay = new HashMap<>();
+
+    DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd"); 
+
+    Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
+    Query<Entity> query = Query.newEntityQueryBuilder().setKind("Comment").build();
+    QueryResults<Entity> results = datastore.run(query);
+
+    while (results.hasNext()) {
```

Comment:
This is always getting the current snapshot of comments stored in the comment datastore. What if someone tried to edit a comment from 4 days ago? Or deleted a comment from a few days ago? Should the comment count for that day suddenly decrease then? 

Instead of getting all the comments and grouping them into day-buckets, why not have a separate datastore that stores the date and comment count? You would have to modify the datastore every time someone adds a new comment or edits a previous comment, but you would be able to keep track of all the activity going on in this website. 

Feel free to address this in a future PR.

Reply:
Hao also recommended that I create a separate datastore for recording the upvotes and downvotes of a given user, so I will do these in the next PR.